## SQL: Сервис для чтения книг по подписке

### Задания

- <a href='#1'>Посчитайте, сколько книг вышло после 1 января 2000 года;</a>
- <a href='#2'>Для каждой книги посчитайте количество обзоров и среднюю оценку;</a>
- <a href='#3'>Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры; </a>
- <a href='#4'>Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками; </a>
- <a href='#5'>Посчитайте среднее количество обзоров от пользователей, которые поставили больше 50 оценок.</a>

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.


In [1]:
import pandas as pd
from sqlalchemy import create_engine


In [2]:
db_config = {'user': 'praktikum_student', # имя пользователя
 'pwd': 'Sdf4$2;d-d30pp', # пароль
 'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
 'port': 6432, # порт подключения
 'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [3]:
def print_table(table_name):
    query = '''
    SELECT 
     *
    FROM books
    '''
    table_name = pd.io.sql.read_sql(query, con = engine)
    display(table_name.head())

In [5]:
books = '''
    SELECT 
     *
    FROM books
    '''
print_table(books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [6]:
authors = '''
    SELECT 
     *
    FROM authors
    '''
print_table(authors)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [7]:
publishers = '''
    SELECT 
     *
    FROM publishers
    '''
print_table(publishers)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [8]:
ratings = '''
    SELECT 
     *
    FROM ratings
    '''
print_table(ratings)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [9]:
reviews = '''
    SELECT 
     *
    FROM reviews
    '''
print_table(reviews)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

<a id='1'></a>  
### Сколько книг вышло после 1 января 2000 года?

In [10]:
query_books = '''
    SELECT 
    COUNT(book_id) as books_count 
    FROM books
    WHERE publication_date :: timestamp > '2000-01-01' :: timestamp'''

In [11]:
books = pd.io.sql.read_sql(query_books, con = engine)
print(books)

   books_count
0          819


После 1 января 200 года вышло 819 книг

<a id='2'></a> 
### Количество обзоров и средняя оценка для каждой книги

In [12]:
query_authors = '''
        SELECT
            books.book_id AS book_id,
            books.title AS book_title,
            SQ2.mean_rating AS mean_rating,
            SQ2.num_rating AS num_rating,
            SQ3.num_review AS num_rewiev
        FROM books
        LEFT JOIN 
        (SELECT 
            ratings.book_id as book_id,
            AVG(ratings.rating) as mean_rating,
            COUNT(ratings.rating) as num_rating
        FROM ratings
        GROUP BY 
            book_id) 
            AS SQ2 ON books.book_id = SQ2.book_id
        LEFT JOIN 
        (SELECT
            reviews.book_id AS book_id,
            COUNT(reviews.review_id) AS num_review
        FROM reviews
        GROUP BY 
            book_id) 
            AS SQ3 ON books.book_id = SQ3.book_id
            '''

In [13]:
books_with_rating = pd.io.sql.read_sql(query_authors, con = engine)
books_with_rating.head()

book_id                                         book_title  mean_rating  \
0      652          The Body in the Library (Miss Marple  #3)     4.500000   
1      273                                          Galápagos     4.500000   
2       51                           A Tree Grows in Brooklyn     4.250000   
3      951  Undaunted Courage: The Pioneering First Missio...     4.000000   
4      839                                        The Prophet     4.285714   

   num_rating  num_rewiev  
0           2         2.0  
1           2         2.0  
2          12         5.0  
3           2         2.0  
4           7         4.0

books_with_rating - таблица, где для каждой книги посчитано количество ревью, количество рейтинговых оценок и средний рейтинг

<a id='3'></a> 
### Какое издательство выпустило наибольшее число книг толще 50 страниц?

In [14]:
query_publishers = '''
    SELECT
        SQ1.publisher AS publisher,
        COUNT(SQ1.book_id) AS books_count
    FROM
        (SELECT 
            books.publisher_id AS publisher_id,
            books.book_id AS book_id,
            books.num_pages AS num_pages,
            publishers.publisher_id,
            publishers.publisher AS publisher
        FROM books
        LEFT JOIN publishers ON books.publisher_id = publishers.publisher_id
        WHERE 
            num_pages > 50) AS SQ1
    GROUP BY
    publisher
    ORDER BY 
    books_count DESC
    LIMIT
    10
    '''
publishers = pd.io.sql.read_sql(query_publishers, con = engine)
publishers.head()

publisher  books_count
0             Penguin Books           42
1                   Vintage           31
2  Grand Central Publishing           25
3          Penguin Classics           24
4          Ballantine Books           19

Издательство, которое выпустило наибольшее число книг толще 50 страниц  - Penguin Books

<a id='4'></a> 
### Автор с самой высокой средней оценкой книг (учитывая только книги с 50 и более оценками)

In [15]:
query_ratings = '''
     SELECT 
     SUB5.author AS author,
     AVG(SUB5.mean_rating) AS mean_rating
     FROM 
     (SELECT
     books.book_id AS book_id,
     authors.author,
     SUB4.mean_rating 
     FROM books
     LEFT JOIN 
            (SELECT 
            ratings.book_id as book_id,
            AVG(ratings.rating) as mean_rating,
            COUNT(ratings.rating) as num_rating
            FROM ratings
            GROUP BY book_id
            HAVING 
            COUNT(ratings.rating) > 50
            ORDER BY 
            mean_rating DESC 
            ) 
            AS SUB4 ON SUB4.book_id = books.book_id
     LEFT JOIN authors ON books.author_id = authors.author_id) AS SUB5
     WHERE 
     mean_rating IS NOT NULL
     GROUP BY 
     author
     ORDER BY
     mean_rating DESC
     LIMIT 
     10
     
            
    '''
ratings = pd.io.sql.read_sql(query_ratings, con = engine)
ratings.head()

author  mean_rating
0         J.K. Rowling/Mary GrandPré     4.283844
1  Markus Zusak/Cao Xuân Việt Khương     4.264151
2                     J.R.R. Tolkien     4.258446
3                  Louisa May Alcott     4.192308
4                       Rick Riordan     4.080645

Автор с самой высокой средней оценкой книг J.K. Rowling/Mary GrandPré

<a id='5'></a>  
### Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [16]:
query_reviews = '''
    SELECT
    AVG(count)
    FROM
    (SELECT
        reviews.username,
        COUNT(reviews.review_id)
    FROM reviews
    WHERE reviews.username IN  
        (SELECT 
            ratings.username AS username
        FROM ratings
        GROUP BY username
        HAVING COUNT(ratings.rating_id) > 50)
    GROUP BY username) AS SUB6
            '''
reviews = pd.io.sql.read_sql(query_reviews, con = engine)
reviews

avg
0  24.333333

Cреднее количество обзоров от пользователей, которые поставили больше 50 оценок = 24